Make code search results scroll to matching fragments

Tunous opened the pull request:
The appearance of code search results was changed so it's now possible to click on individual matched code blocks. When doing so the opened file will be automatically scrolled to the matched block.

![peek 2017-05-24 14-55](https://cloud.githubusercontent.com/assets/5156340/26404191/41984ee0-4091-11e7-9794-1a964eb47fb2.gif)

Closes #597.

maniac103 commented on app/src/main/java/com/gh4a/utils/StringUtils.java:
```diff
@@ -139,4 +141,30 @@ public static SpannableStringBuilder applyBoldTags(Context context,
+            }
+
+            if (isMatch) {
+                return new Pair<>(i + 1, i + matchLines.length);
```
Return an int[2] here?

maniac103 commented on app/src/main/java/com/gh4a/utils/StringUtils.java:
```diff
@@ -139,4 +141,30 @@ public static SpannableStringBuilder applyBoldTags(Context context,
+                String fLine = matchLines[j];
+                if (j == 0 && !cLine.endsWith(fLine) ||
+                        j != 0 && j != matchLines.length - 1 && !cLine.equals(fLine) ||
+                        j == matchLines.length - 1 && !cLine.startsWith(fLine)) {
```
Suggest parenthesis around the && statements.

maniac103 commented on app/src/main/java/com/gh4a/adapter/CodeSearchAdapter.java:
```diff
@@ -55,28 +59,48 @@ protected void onBindViewHolder(ViewHolder holder, CodeSearchResult result) {
                         int start = item.getStartPos();
                         int end = item.getEdndPos();
```
Now that I see this, we should probably fix this :)

Tunous replied:
What do you mean?

maniac103 replied:
getEdndPos -> getEndPos ;)

Tunous replied:
Ohh I see now. Also, I'm not sure if you are using the app right now, but if you do then replying looks to be broken: 
![spectacle x14496](https://cloud.githubusercontent.com/assets/5156340/26406358/5ca07b2a-4098-11e7-8dae-dd036040769e.png)

Tunous commented:
Updated.

maniac103 commented on app/src/main/java/com/gh4a/utils/StringUtils.java:
```diff
@@ -139,4 +140,30 @@ public static SpannableStringBuilder applyBoldTags(Context context,
+        String[] lines = input.split("\n");
+        String[] matchLines = match.split("\n");
+
+        for (int i = 0; i < lines.length - matchLines.length; i++) {
```
There's one issue in this loop: If input and match are exactly equal, the break condition in this line will cause no match to be found.

maniac103 commented on app/src/main/java/com/gh4a/utils/StringUtils.java:
```diff
@@ -139,4 +140,30 @@ public static SpannableStringBuilder applyBoldTags(Context context,
+                String fLine = matchLines[j];
+                if ((j == 0 && !cLine.endsWith(fLine)) ||
+                        (j != 0 && j != matchLines.length - 1 && !cLine.equals(fLine)) ||
+                        (j == matchLines.length - 1 && !cLine.startsWith(fLine))) {
```
Wouldn't this be easier to read:

```
boolean lineMatches =
        j == 0 ? cLine.endsWith(fLine) :
        j == matchLines.length - 1 ? cLine.startsWith(fLine) :
        cLine.equals(fLine);
if (!lineMatches) {
    isMatch = false;
    break;
}
```

Another naming suggestion: cLine -> inputLine, fLine -> matchLine

maniac103 commented on app/src/main/java/com/gh4a/utils/StringUtils.java:
```diff
@@ -139,4 +140,30 @@ public static SpannableStringBuilder applyBoldTags(Context context,
+            }
+
+            if (isMatch) {
+                return new int[] { i + 1, i + matchLines.length };
```
Shouldn't this be `{i, i + matchLines.length - 1}`? A match in the first line currently yields {1, 1}, not {0, 0}.

maniac103 commented:
Maybe an even simpler matching algorithm could be this (untested):

```
int pos = input.indexOf(match);
if (pos < 0) {
    return null;
}
int total = 0, count = 0;
for (String line : lines.split("\n")) {
    if (pos >= total && pos < total + line.length()) {
        return new int[] { count, count + match.split("\n").length };
    }
    total += line.length;
    count++;
}
```

Doing so would have the advantage of having only a single loop, not two nested ones.

Tunous commented:
It's a bit off. What about this:

```java
int pos = input.indexOf(match);
if (pos < 0) {
    return null;
}

int start = input.substring(0, pos).split("\n").length;
int end = start + match.split("\n").length - 1;
return new int[] { start, end };
```

Looks to be working so far.

maniac103 commented:
Yeah, this one seems to be working.
